### PR TITLE
Fix pop-in issues

### PR DESCRIPTION
**Pull request by Oliveriver**
**Brief Description of What This PR Does**

Fixes entities spawning too close to the player, causing visible pop-in.

**Related Issues**

Fixes https://github.com/Revolutionary-Games/Thrive/issues/3441

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by Oliveriver** [comment]
> That's how it has always been. And the entity count is fully intended to only have entities in it.

That's my point. Without this fix, clouds (specifically those in queued spawns) are being added to the entity count.

File: src/microbe_stage/SpawnSystem.cs [hunk 1]
```diff
@@ -224,7 +226,9 @@ private void HandleQueuedSpawns(ref int spawnsLeftThisFrame)
                     enumerator.Current ?? throw new Exception("Queued spawn enumerator returned null"),
                     spawn.SpawnType);
 
-                ++estimateEntityCount;
+                if (!spawningClouds)
```
[the remark this answers is not in the document]
Comment (Oliveriver) [reply src/microbe_stage/SpawnSystem.cs:229]: Ok, I misunderstood how spawning clouds worked. You're right, it's not possible to reach this method with a cloud spawn type. I'll remove that "fix" then.

File: src/microbe_stage/SpawnSystem.cs [hunk 2]
```diff
@@ -207,10 +207,12 @@ private void HandleQueuedSpawns(ref int spawnsLeftThisFrame)
         {
             var spawn = queuedSpawns.First();
             var enumerator = spawn.Spawns;
+            bool spawningClouds = spawn.SpawnType is CompoundCloudSpawner;
 
-            bool finished = false;
+            // Discard the whole spawn if we're too close to the player
+            bool finished = (playerPosition - spawn.Location).Length() < Constants.SPAWN_SECTOR_SIZE;
```
[the remark this answers is not in the document]
Comment (Oliveriver) [reply src/microbe_stage/SpawnSystem.cs:213]: I would prefer that too, but since the enumerator spawns each time it's iterated, there's no way to check that here before the thing gets spawned. It would have to be a check in the spawn method of every spawn type, which would require passing the player location to each of these spawn types and doing this comparison for every single spawn, not just the queued ones. At that point I feel it's just too much of a mess to be worth it.

Alternatively, it could despawn straight away after being spawned, but I can't see that ending well performance-wise, and that might even still be visible to the player. The other alternative is to overhaul how the spawn system enumerates stuff, which I'm quite against considering all the extra design and testing that would require.

Basically, I don't see any other way of doing this that's anywhere near this simple. Especially since as far as I can tell, location here will always be the correct location anyway. But I'm willing to believe I've missed something vital, so feel free to correct me.

[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/microbe_stage/SpawnSystem.cs:213]: > I would prefer that too, but since the enumerator spawns each time it's iterated, there's no way to check that here before the thing gets spawned. It would have to be a check in the spawn method of every spawn type, which would require passing the player location to each of these spawn types and doing this comparison for every single spawn, not just the queued ones. At that point I feel it's just too much of a mess to be worth it.

I said earlier that it's *probably* not too much of a problem to just despawn the entity immediately.
Anything that happens during a single frame should not be able to be visible.

> location here will always be the correct location anyway.

Your approach anyway is not actually bulletproof either because the spawners don't use the exact position they are given, so depending on that to skip spawning has situations where someone might write a badly behaving spawner or something that causes the approach to break.

**Review by hhyyrylainen — APPROVED** [review]
Code looks fine and I tested this a bit on the default entity limit to make sure this doesn't constantly cause things that couldn't spawn in the end. So to me this looks fine, but let's wait a bit more to see if other testing gets done.

**Comment by hhyyrylainen** [comment]
I see testing just got done on this, so I guess we'll merge this now.
